apps/src/p5lab/spritelab/commands/validationCommands.js

```diff
@@ -57,10 +64,583 @@ export const commands = {
+    this.validationTimes.wait = frames;
+  },
+  setDelayTime(frames) {
+    this.validationTimes.delay = frames;
+  },
+
+  // Used in levels (typically every frame) to validate based on all criteria
+  validate() {
+    // Get the current (ie. previous frame) pass/fail state prior to validation
+    const state = commands.getPassState(this.criteria);
+
+    const barWidth =
+      this.currentFrame() * commands.calculateBarScale(this.validationTimes);
+    drawUtils.validationBar(this.p5, barWidth, state, {});
+
+    // Check all criteria and update the completion status of each.
+    if (this.currentFrame() <= this.validationTimes.wait) {
+      commands.checkAllCriteria(this.criteria);
+    } else {
+      // If "wait time" is over, determine if student passes or fails.
+      var results = {};
+      if (state === 'fail') {
+        console.log(this.criteria);
+        results = {
+          state: 'failed',
+          feedback: commands.reportFailure(this.criteria)
+        };
+      } else if (
+        this.currentFrame() >=
+        this.validationTimes.wait + this.validationTimes.delay
+      ) {
+        results = {
+          state: 'succeeded',
+          feedback: commands.reportSuccess(this.criteria)
+        };
+      }
+      return results;
+    }
+    commands.updatePrevious.call(this);
+  },
+
+  // In levels with events, we need to check if things happened over time.
+  // Store information about this frame for comparisons during the next frame.
+  updatePrevious() {
+    const spriteIds = this.getSpriteIdsInUse();
+    this.previous.eventLogLength = this.eventLog.length;
+    this.previous.printLogLength = this.printLog.length || 0;
+    this.previous.sprites = [];
+    for (let i = 0; i < spriteIds.length; i++) {
+      let spriteId = spriteIds[i];
+      this.previous.sprites.push({
+        id: spriteId,
+        costume: this.nativeSpriteMap[spriteId].getAnimationLabel(),
+        x: this.nativeSpriteMap[spriteId].x,
+        y: this.nativeSpriteMap[spriteId].y,
+        behaviors: this.getBehaviorsForSpriteId(spriteId),
+        tint: this.nativeSpriteMap[spriteId].tint || '',
+        scale: this.nativeSpriteMap[spriteId].getScale(),
+        speed: this.nativeSpriteMap[spriteId].speed,
+        rotation: this.nativeSpriteMap[spriteId].rotation
+      });
+    }
+  },
+
+  // If the student has not completed any criteria, they are "failing".
+  // This determines things like the current color of the validation timer bar below the play area.
+  getPassState(criteria) {
+    var state = 'pass';
+    for (const criterion in criteria) {
+      if (!criteria[criterion].complete) {
+        state = 'fail';
+      }
+    }
+    return state;
+  },
+
+  calculateBarScale(validationTimes) {
+    return 400 / (validationTimes.wait + validationTimes.delay);
+  },
+
+  checkAllCriteria(criteria) {
+    for (const criterion in criteria) {
+      if (!criteria[criterion].complete) {
+        if (criteria[criterion].predicate()) {
```

Comment:
As mentioned before, as a newcomer to this code I'm not sure what a "predicate" is.. a comment in the file could help get bearings.